src/NumericalAlgorithms/SphericalHarmonic/ComplexDataSlice.hpp

```diff
@@ -0,0 +1,171 @@
+  /// `size` elements. Need not be a part of a ComplexDataVector.
+  ///
+  /// \details A copy occurs in the case of a `real_slices` kind of slice.
+  ComplexDataSlice(std::complex<double>* start, size_t size) noexcept;
+
+  // For the lifetime of the data slice, it points to the same portion of a
+  // single vector, default move assign behavior would contradict that
+  ComplexDataSlice() = delete;
+  ComplexDataSlice(ComplexDataSlice&) = default;
+  ComplexDataSlice(ComplexDataSlice&&) = default;
+  ComplexDataSlice operator=(ComplexDataSlice&&) = delete;
+  ~ComplexDataSlice() = default;
+
+  /// \brief Assign into the slice the values of a same-sized ComplexDataVector
+  ///
+  /// \details For either type of slice, this causes a copy from the provided
+  /// vector. For a `real_slices`, it is effectively two copies (one for each
+  /// component), but acts that way to use faster vector assignments.
+  ComplexDataSlice<Kind>& operator=(const ComplexDataVector& vector) noexcept;
+
+  /// \brief Assign into the slice the values from a same-sized `interleaved`
+  /// slice
+  ///
+  /// \details For either type of slice, this causes a copy. For a
+  /// `real_slices`, it is effectively two copies (one for each component), but
+  /// acts that way to use faster vector assignments.
+  // clang-tidy: non-const reference; this is a copy constructor which needs
+  // to access the data in another slice, which can force an internal copy.
+  ComplexDataSlice<Kind>& operator=(
+      ComplexDataSlice<ComplexSliceKind::interleaved>&
+          slice) noexcept;  // NOLINT
+
+  /// \brief Assign into the slice the values from a same-sized `real_slices`
+  /// slice
+  ///
+  /// \details For either type of slice, this causes a copy, but avoids a second
+  /// copy in the provided slice.
+  // clang-tidy: non-const reference; this is a copy constructor which needs
+  // to access the data in another slice, which can force an internal copy.
+  ComplexDataSlice<Kind>& operator=(
+      ComplexDataSlice<ComplexSliceKind::real_slices>&
+          slice) noexcept;  // NOLINT
+
+  /// \brief Assign into the real components of a slice the values from a
+  /// provided `DataVector`
+  ///
+  /// \details For either type of slice, a copy must occur. For `real_slices`
+  /// the copy is a contiguous vector copy, while for `interleaved` slices, the
+  /// copy must distribute the values to alternating memory locations.
+  ComplexDataSlice<Kind>& assign_real(const DataVector& vector) noexcept;
+
+  /// \brief Assign into the imaginary components of a slice the values from a
+  /// provided `DataVector`
+  ///
+  /// \details For either type of slice, a copy must occur. For `real_slices`
+  /// the copy is a contiguous vector copy, while for `interleaved` slices, the
+  /// copy must distribute the values to alternating memory locations.
+  ComplexDataSlice<Kind>& assign_imag(const DataVector& vector) noexcept;
```

Comment:
Don't need to specify the template parameter on the return type of all of these.